pyzo/codeeditor/extensions/appearance.py

```diff
@@ -108,14 +109,64 @@ def paintEvent(self, event):
         super(HighlightMatchingOccurrences, self).paintEvent(event)
 
 
+class _ParenNotFound(Exception) :
+    pass
+
+class _ParenIterator :
+    """ Iterates in given direction) over parentheses in the document.
+    Uses the stored token-list of the blocks.
+    Iteration gives both a ParenthesisToken and the global position of
+    the block it lies in (because positions in the ParenthesisToken is
+    relative to the block)."""
+    def __init__(self, cursor, direction) :
+        self.cur_block = cursor.block()
+        self.cur_tokens = self._getParenTokens()
+        self.direction = direction
+        # We need to know where we start in the current token list
+        k = 0
+        try :
+            while self.cur_tokens[k].end != cursor.positionInBlock() :
+                k += 1
+            self.cur_pos = k
+        except IndexError :
+            # If the parenthesis cannot be found, it means that it is not inluded
+            # in any token, ie. it is part of a string or comment
+            raise _ParenNotFound
+    
+    def _getParenTokens(self) :
+        return list(filter(lambda x : isinstance(x, ParenthesisToken), self.cur_block.userData().tokens))
```

Comment:
Better test for tokens is None. The tokens attribute always exist, because I modified the __init__ of userData to set it to None bye default (being used to statically typed languages like OCaml, I am always very ill at ease when functions just add attributes to objects that happen to pass in their hands ; it also took me several hours to understand that the parser just adds a _style attribute to tokens when it sees fit to do so).